src/components/templates/PosterHall/PosterHall.tsx

```diff
@@ -39,8 +47,52 @@ export const PosterHall: React.FC<PosterHallProps> = ({ venue }) => {
     ));
   }, [posterVenues]);
 
+  const [liveVenueIds, setLiveVenueIds] = useState<string[]>();
+
+  const { relatedVenues } = useRelatedVenues({
+    currentVenueId: venue.id,
+  });
+
+  const subVenues = useMemo(() => {
+    return relatedVenues.filter(
+      (relatedVenue) =>
+        relatedVenue.parentId === venue.id &&
+        relatedVenue.template !== VenueTemplate.posterpage
+    );
+  }, [relatedVenues, venue]);
+
+  const subVenueIds = useMemo(() => subVenues.map((venue) => venue.id), [
+    subVenues,
+  ]);
+
+  const { events: subVenueEvents } = useVenueEvents({
+    venueIds: subVenueIds,
+  });
+
+  useInterval(() => {
+    setLiveVenueIds(
+      subVenueEvents
+        .filter((event) => isEventLive(event))
+        .map((event) => event.venueId)
+    );
+  }, POSTERHALL_SUBVENUE_STATUS_MS);
```

Comment:
This will cause fetching of the events every second

Reply:
`subVenueEvents` is loaded from `useVenueEvents`, which unless it has changed since I last looked, will only load the events when the `venueIds` passed to it changes. In this case, that comes from the `useMemo`'d `subVenueIds`, which comes from the `useMemo`'d `subVenues`, which filters `relatedVenues`, which comes from `useRelatedVenues` which will only break memo/change when `sovereignVenueId` changes (so basically never).

So every second (or whatever `POSTERHALL_SUBVENUE_STATUS_MS` equates to), this `useInterval` will run and call `setLiveVenueIds`, which I believe (not 100% sure) will cause a re-render, even if the values end up being the same. Since `liveVenueIds` will have changed, it will break the memo on `renderedSubvenues`.

If you use the 'functional' form of setState (eg. `setLiveVenueIds(prevState => newState`), and then check if the values actually changed, you can return the `prevState` when they haven't, which will skip re-renders as per:

- https://reactjs.org/docs/hooks-reference.html#usestate
  - > If the new state is computed using the previous state, you can pass a function to `setState`. The function will receive the previous value, and return an updated value.
  - > **If your update function returns the exact same value as the current state, the subsequent rerender will be skipped completely.**

Reply:
In short, I don't think the situation is as dire as @mike-lvov originally suspected it to be, but there are definitely a number of optimisations that could be made, as spelled out above.

I haven't looked at the changes @margulies made in his most recent changed mentioned in https://github.com/sparkletown/sparkle/pull/1432/files/189959cad6f72e004d8186d16b63876952d17c8c#r649509502 since the hash wasn't linked for it. Generally speaking reducing the frequency of the `useInterval` to 1min will reduce the impact of any re-renders/etc.

Reply:
I've implemented a change in this commit: c14d9678647777019fa4f36de1482747a532ab75
that only calls `setLiveVenueIds` inside `useInterval` if there is a change in the `liveVenueIds` list

If there's another way you'd prefer that I implement that, just let me know

Reply:
This has now been moved to `hook/poster.ts`, implemented in a similar way to the calculation of `currentTimePosition` in `components/molecules/Schedule/Schedule.tsx`. Commit: 2f5bb2560106081d10ef3eb1f733c32207b59406

Reply:
Looking at your original commit there, the separate useMemo for the unit code init  needed. If you make the initial value of the useState a function, it will only be run the first time it’s needed to set the initial state.

Then the === check to see if it should update won’t work because it will be checking that the 2 arrays have the literal same reference, not that they contain the same items.

Reply:
In the 2nd commit.

checkLiveEvents should be named better for what it’s doing, maybe filterLiveEvents or similar.

The variable subVenues is somewhat misleading, as it’s really representing subVenuesWithoutPosterPages.

setLiveEvents would probably be better named updateLiveEvents. Also, within it you are needlessly running the filter multiple times. Run it once and store it in a const, then use that in your checks. You still have the same === bug as mentioned above that will check that they have the same literal array reference, not that the items in the array are the same. I’ll often use JavaScript’s Set type when I need to efficiently perform this sort of check, but unsure of the overhead of converting both arrays to sets every time just for the comparison here, probably an easier way to check they have all values. Probably a lodash function that does it. You also need a newline after your guard ‘if’.

Whats the point of the useEffect that calls setLiveEvents? You already run the equivalent code in the initialiser of the useState.

Then once that is removed, since you only use setLiveEvents in the useInterval, there doesn’t seem to be a need for it to be extracted as a useCallback anyway, so can just be inclined into the useInterval

Reply:
Thanks for explaining these points. Several of them are addressed in the current version. I've also removed the `if` line from `useEffect` now. 

I'll rename `checkLiveEvents` and `subVenue` now as well.

Reply:
I've renamed `subVenue` -> `nonPosterSubVenue` (including variants such as `liveNonPosterSubVenueEvents`)

Reply:
> I've also removed the if line from useEffect now.

@margulies Why did you remove it?

Reply:
Sorry, that was actually a type -- I'd removed the `if` statement from `useInterval`. As you'd mentioned initially, I assumed that `setLiveNonPosterSubVenueEvents(() => filterLiveEvents(nonPosterSubVenueEvents));` would only update if there was a change. Let me know if I should bring back the `if` statement though to prevent this from running if there's no change.

Reply:
> Let me know if I should bring back the if statement though to prevent this from running if there's no change.

Definitely will need to do this.

---

From https://github.com/sparkletown/sparkle/pull/1432#discussion_r649631983:

Particularly note the bolded part:

> If you use the 'functional' form of setState (eg. setLiveVenueIds(prevState => newState), **and then check if the values actually changed, you can return the prevState when they haven't, which will skip re-renders**

From https://github.com/sparkletown/sparkle/pull/1432#discussion_r650445196:

> You still have the same === bug as mentioned above that will check that they have the same literal array reference, not that the items in the array are the same. I’ll often use JavaScript’s Set type when I need to efficiently perform this sort of check, but unsure of the overhead of converting both arrays to sets every time just for the comparison here, probably an easier way to check they have all values. Probably a lodash function that does it. You also need a newline after your guard ‘if’.

Reply:
Thanks! I've now implemented the following `if` guard in 4c6a772dfac90f498a767199c71964886ab7740e
I'm unclear if this properly addresses the first point you raised regarding the functional form of `setState`, so if that needs to change as well, please let me know.

```
useInterval(() => {
    if (
      isEqual(
        sortBy(filterLiveEvents(nonPosterSubVenueEvents)),
        sortBy(liveNonPosterSubVenueEvents)
      )
    )
      return;

    setLiveNonPosterSubVenueEvents(() =>
      filterLiveEvents(nonPosterSubVenueEvents)
    );
  }, POSTERHALL_SUBVENUE_STATUS_MS);
```

Reply:
This has the same problem of running the filter code multiple times that I said to extract it as a const and reuse that.

Do you need to sort them? Is there no helper that will directly check that one array contains all the elements of the other and vice versa?

Having the guard where it is is going to close over the value of liveNonPosterVenues at the time the useInterval is created, so it’s not going to be updated. That’s why you need to put the guard within the functional form of the  set* function. When they are the same, return the prevState. When they aren’t, return the new state as I originally described.

Reply:
Sorry for being a bit dense on this one. I've refactored as follows in 1a8356b3d62b2e0d19b285bb64737b00ecfae908 

Here's my understanding:
```
  // const that will be called to pull an object containing only live events:

  const filteredLiveEvents = filterLiveEvents(nonPosterSubVenueEvents);

  // set up state type -- I may have already misunderstood -- is this where 
  // the functional form below should already be created?:

  const [
    liveNonPosterSubVenueEvents,
    setLiveNonPosterSubVenueEvents,
  ] = useState<WithVenueId<VenueEvent>[]>();

  // run set... here to ensure it runs after the consts have all loaded 
  // and doesn't wait for the first useInterval timer

  useEffect(
    () =>
      setLiveNonPosterSubVenueEvents(filteredLiveEvents),
    [filteredLiveEvents]
  );

  // check every 1 min if there's a change from prevState using lodash `isEqual`. 
  // If they're equal, return preState (so no reloading). If they aren't equal, return the 
  // new state, initiating a reloading.

  useInterval(() => {
    setLiveNonPosterSubVenueEvents(function(prevState){
      if (isEqual(prevState,filteredLiveEvents)) return prevState;

      return filteredLiveEvents;
    });
  }, POSTERHALL_SUBVENUE_STATUS_MS);
```

Reply:
Or is this the correct syntax?: 
```
useInterval(() => {
    setLiveNonPosterSubVenueEvents(prevState => {
      if (isEqual(prevState, filteredLiveEvents)) return prevState;

      return filteredLiveEvents;
    });
  }, POSTERHALL_SUBVENUE_STATUS_MS);
```

Reply:
One last attempt, and then I promise to be patient. Is this any closer to what we're going for: 
```
const [
    liveNonPosterSubVenueEvents,
    setLiveNonPosterSubVenueEvents,
  ] = useState<WithVenueId<VenueEvent>[]>();

  const updateLiveEvents = useCallback(() => {
    const filteredLiveEvents = filterLiveEvents(nonPosterSubVenueEvents);

    setLiveNonPosterSubVenueEvents((prevState) => {
      if (isEqual(prevState, filteredLiveEvents)) return prevState;

      return filteredLiveEvents;
    })
  },[nonPosterSubVenueEvents]);

  useEffect(() => updateLiveEvents());

  useInterval(() => {
    updateLiveEvents();
  }, POSTERHALL_SUBVENUE_STATUS_MS);
```

Reply:
I've checked the refresh rate on a version similar to the one above, and it seems to be working now: fd09688df03497142bbd53fd069876cc53995f3e Of course, let me know if any changes should be made, but hope this is on the right track